javalib/src/main/scala/java/util/regex/GroupStartMapper.scala

```diff
@@ -0,0 +1,497 @@
+     *
+     *  @return 1 plus the largest assigned group number.
+     */
+    def setNewGroup(newGroupIndex: Int): Int = {
+      newGroup = newGroupIndex
+      newGroupIndex + 1
+    }
+
+    def buildRegex(groupNodeMap: js.Array[Node]): String
+
+    /* When assigning group positions. I could not choose between assigning
+     * group numbers from left to right or from right to left, because there
+     * both failed in one case each. Normally, both directions give the same
+     * result. But there are corner cases.
+     *
+     * Consider the following regex matching `abbbbbbc`
+     *
+     *     (?=ab*(c))ab
+     *
+     * After conversion, this becomes:
+     *
+     *     (?=(ab*)(c))(ab)
+     *
+     * To recover the position of the group (c), we cannot do anything but
+     * compute it from the length of (ab*), that is, propagate the start,
+     * compute the length, and return the end, and this, recursively. This is
+     * what we need to do in a forward-matching regex.
+     *
+     * However, consider the following regex matching `abcbdbe`
+     *
+     *     a(b.)*
+     *
+     * After conversion, it is transformed to:
+     *
+     *     (a)((b.)*)
+     *
+     * The semantics of group matching under a star are that the last match is
+     * kept. Therefore, we cannot obtain the start position of (b.) from
+     * propagating lengths from left to right. What we first do is to get the
+     * start, then the end, of the group `((b.)*)`, and then we propagate this
+     * end to the inner group.
+     *
+     * Note that when javascript will support back-matching `(?<= )` and
+     * `(?<! )` (hopefully one day), we will be able to implement the length
+     * transfer using the `setEndReturnStart` method, because we have no clue
+     * on where the match started (this is different from the `start` position
+     * because it can extend before it).
+     *
+     * Zero-length test nodes of the type `(?= )` or `(?! )` do not
+     * count to propagate the length on the right or on the left.
+     *
+     * Every group having a *repeater* uses the semantics of propagating the
+     * end to the start.
+     */
+
+    /** Propagates the start or end position of this node to its descendants.
+     *
+     *  According to the big comment above, `RepeatedNode`s propagate the
+     *  `end`, while other nodes propagate the `start`.
+     */
+    def propagate(matchResult: js.RegExp.ExecResult,
+        groupStartMap: js.Array[Int], start: Int, end: Int): Unit
+
+    def setEndReturnStart(matchResult: js.RegExp.ExecResult,
+        groupStartMap: js.Array[Int], end: Int): Int = {
+
+      val start = matchResult(newGroup).fold {
+        -1
+      } { matched =>
+        val start = end - matched.length
+        if (this.isInstanceOf[GroupNode])
+          groupStartMap(this.asInstanceOf[GroupNode].number) = start
+        start
+      }
+      propagate(matchResult, groupStartMap, start, end)
```

Comment:
What happens if we propagate a -1 here?

Reply:
TBH I'm not quite sure. It's doing whatever it did before this PR. I believe it doesn't really matter; at least it doesn't seem to matter in any of our tests.